### DISCUSSION TITLE
Help adding back the '-add-pdf-to-library' functionality

### DISCUSSION THREAD

**Discussion by apc** [emacs-citar/citar#522] [Q&A] [answered]
First: thanks, again, for the work you all continue to do on this!

Second: I'm trying to add back the functionality from some previous version that allowed me to add PDFs to my library using citar/bibtex-actions. I realize that functionality in turn depended on `bibtex-completion`, so I just modified the function from `bibtex-completion` and added it to my config. 

My question, now, is what the best way of adding this function to the list of actions made available via Embark. Say the function is called `+citar-add-pdf-to-library`. I then added that a binding to that function in `citar-map`, and ended up with something like this in my `use-package` declaration: 

````el
(use-package citar
  :bind (("C-c b" . citar-insert-citation)
         :map citar-map
         ("a" . +citar-add-pdf-to-library))
  :after embark
  :init
  :custom
  (citar-bibliography '("~/bib/apcmain.bib"))
  (citar-library-paths '("~/Documents/Papers"))
  (citar-notes-paths '("~/org/notebook"))
  (org-cite-global-bibliography '("~/bib/apcmain.bib"))
  (org-cite-insert-processor 'citar)
  (org-cite-follow-processor 'citar)
  (org-cite-activate-processor 'citar)
  (citar-at-point-function 'embark-act)
  (citar-templates
      '((main . "${author editor:15}   ${title:40}   ${date year issued:4}")
        (suffix . "          ${=key= id:15}    ${=type=:12}    ${tags keywords:*}")
        (note . "Notes on ${author editor}, ${title}")))
  :config
  (defvar apc-bibtex-pdf-extension ".pdf")
  (defun +citar-add-pdf-to-library (keys)
  "Add a PDF to the library for the first entry in KEYS.
The PDF can be added either from an open buffer, a file, or a
URL."
  (interactive)
  (let* ((key (car keys))
         (source (char-to-string
                  (read-char-choice "Add pdf from [b]uffer, [f]ile, or [u]rl? " '(?b ?f ?u))))
         (buffer (when (string= source "b")
                   (read-buffer-to-switch "Add pdf buffer: ")))
         (file (when (string= source "f")
                 (expand-file-name (read-file-name "Add pdf file: " nil nil t))))
         (url (when (string= source "u")
                (read-string "Add pdf URL: ")))
         (path (-flatten (list citar-library-paths)))
         (path (if (cdr path)
                   (completing-read "Add pdf to: " path nil t)
                 (car path)))
         (pdf (expand-file-name (completing-read "Rename pdf to: "
                                                 (--map (s-concat key it)
                                                        (-flatten apc-bibtex-pdf-extension))
                                                 nil nil key)
                                path)))
    (cond
     (buffer
      (with-current-buffer buffer
        (write-file pdf t)))
     (file
      (copy-file file pdf 1))
     (url
      (url-copy-file url pdf 1)))))
````

Now, the function appears when I call `embark-act` from my list of citations, but calling the function gives me an error, starting with

```
embark PCH: (wrong-number-of-arguments ((t) (keys) "Add a PDF to the library for the first entry in KEYS.
The PDF can be added either from an open buffer, a file, or a
URL."
```

Clearly I'm missing something about how Embark works, but since at some point you had this functionality as part of `citar`, I'm hoping you might be able to help.

**Comment by bdarcus** [comment]
What's your preferred UX workflow for the basic command; from the minibuffer?

**Reply by apc** [reply]
Yeah, I guess so.

**Reply by bdarcus** [reply]
Sorry; I poorly worded my question.

Reread the first part, and ignore the last clause.

**Reply by apc** [reply]
Hm. I guess I'm not really sure I understand the question. I'd like to be able to call e.g. `citar-insert-citation` and, after selecting one of the entries in the minibuffer, call `embark-act` and be prompted with the option of adding a PDF file to the library and naming it using the citation key for the selected entry. Does that help?

**Comment by apc** [comment, marked as answer]
OK, I think I figured out what was missing. This seems to work now: 

```el

(use-package citar
  :bind (("C-c b" . citar-insert-citation)
         :map minibuffer-local-map
         ("M-b" . citar-insert-preset)
         :map citar-map
         ("a" . +citar-add-pdf-to-library))
  :after embark
  :init
  :custom
  (citar-bibliography '("~/bib/apcmain.bib"))
  (citar-library-paths '("~/Documents/Papers"))
  (citar-notes-paths '("~/org/notebook"))
  (org-cite-global-bibliography '("~/bib/apcmain.bib"))
  (org-cite-insert-processor 'citar)
  (org-cite-follow-processor 'citar)
  (org-cite-activate-processor 'citar)
  (citar-at-point-function 'embark-act)
  (citar-templates
      '((main . "${author editor:15}   ${title:40}   ${date year issued:4}")
        (suffix . "          ${=key= id:15}    ${=type=:12}    ${tags keywords:*}")
        (note . "Notes on ${author editor}, ${title}")))
  :config
  (defvar apc-bibtex-pdf-extension ".pdf")
  (defun +bibtex-add-pdf-to-library (keys)
  "Add a PDF to the library for the first entry in KEYS.
The PDF can be added either from an open buffer, a file, or a
URL."
  (interactive)
  (let* ((key (car keys))
         (source (char-to-string
                  (read-char-choice "Add pdf from [b]uffer, [f]ile, or [u]rl? " '(?b ?f ?u))))
         (buffer (when (string= source "b")
                   (read-buffer-to-switch "Add pdf buffer: ")))
         (file (when (string= source "f")
                 (expand-file-name (read-file-name "Add pdf file: " nil nil t))))
         (url (when (string= source "u")
                (read-string "Add pdf URL: ")))
         (path (-flatten (list citar-library-paths)))
         (path (if (cdr path)
                   (completing-read "Add pdf to: " path nil t)
                 (car path)))
         (pdf (expand-file-name (completing-read "Rename pdf to: "
                                                 (--map (s-concat key it)
                                                        (-flatten apc-bibtex-pdf-extension))
                                                 nil nil key)
                                path)))
    (cond
     (buffer
      (with-current-buffer buffer
        (write-file pdf t)))
     (file
      (copy-file file pdf 1))
     (url
      (url-copy-file url pdf 1)))))
  (defun +citar-add-pdf-to-library (keys-entries)
  "Add PDF associated with the KEYS-ENTRIES to library.
The PDF can be added either from an open buffer, a file, or a
URL.
With prefix, rebuild the cache before offering candidates."
  (interactive (list (citar-select-refs
                      :rebuild-cache current-prefix-arg)))
   (+bibtex-add-pdf-to-library
    (citar--extract-keys keys-entries))))

```

But I leave this open in case you have any suggestions for how to improve on what I've got so far. Thanks again!

**Reply by bdarcus** [reply]
I added a modified, more general, version of this to the wiki. I may add it back to citar; I just wasn't sure if anyone used this ;-)

https://github.com/bdarcus/citar/wiki/Example-functions#add-file-to-library

It uses `completing-read` rather than `completing-read-multiple`, and works on any file.

Note: recent changes in many of these commands are designed to make them compatible with `embark-act-all`. I'm not sure the UI on the current version of this command actually is, seems it seems to prompt you to rename the file in one case. Worth testing this.

It also occurs to me that we might want to include an option to add from a DOI, which maybe does that without prompting if the entry has a DOI?

There's still a little bug in it that I'll fix.

**Reply by apc** [reply]
Yes. Adding from doi would be great!

**Reply by apc** [reply]
Oh, and is the bug related to this error message? 

```
embark PCH: (wrong-type-argument characterp "test2021")
```

**Reply by bdarcus** [reply]
I don't believe so, though I have some other embark-related issue in my local doom install.

**Reply by dschrempf** [reply]
I just wanted to add my support for adding functionality to manage files with `citar`. At the moment, `citar` can view/index associated files such as PDFs but not manage (add or remove) them, is this correct?

**Reply by dschrempf** [reply]
Just to add some thoughts: I could also make good use of a function "Check if there are file (PDF) orphans in the library". Or "The key you just renamed has a PDF file attached. Do you also want to rename the PDF?". (Actually this should probably also be done for notes).

**Reply by bdarcus** [reply]
> I just wanted to add my support for adding functionality to manage files with `citar`. At the moment, `citar` can view/index associated files such as PDFs but not manage (add or remove) them, is this correct?

Correct; it's not been a priority for me as I use Zotero and Better BibTeX to manage this stuff ATM, and nobody else has brought it up.

> I could also make good use of a function "Check if there are file (PDF) orphans in the library". Or "The key you just renamed has a PDF file attached. Do you also want to rename the PDF?". (Actually this should probably also be done for notes).

If you like, create an issue (though it may be awhile before someone gets to it)?

I haven't touched that add-file-to-library function since I posted it on the wiki, but it would make sense to include that I suppose.

**Reply by bdarcus** [reply]
BTW, @dschrempf, #562 relates to your request, I think, but I haven't thought about it in broader context.

**Reply by dschrempf** [reply]
That's right! Although, funnily enough, I renamed your emplate to `...-add-pdf-to-library` because that's the only thing I really need. But a more general function is always nice.
